usdigitalresponse/usdr-gost
# 938 consume grants from sqs

### Ticket #938 
## Description

This PR adds a new library and associated script for consuming SQS messages originating from events published by the `grants-ingest` service. 

It's worth stating up-front that due to forthcoming schema design definitions in https://github.com/usdigitalresponse/grants-ingest/issues/114 (and subsequent incorporation into the published events), the new `grants-ingest` library code will ~likely~ definitely need to be updated soon – specifically, the logic involved in mapping data from the `grants-ingest`-provided format to the format used in the GOST `grants` table. However, given that I think everything else in this branch is here to stay, along with the fact that we won't be running the script in Production immediately, I it would be preferable to get these changes merged in sooner rather than later and avoid maintaining a long-running branch.

Summary of changes:
- Adds a new library module at `packages/server/src/lib/grants-ingest.js` that exports two functions:
  - `receiveNextMessageBatch()`: Long-polls an SQS queue for messages containing data for recently-updated grant opportunities. These messages will be delivered to the queue from events emitted by the `grants-ingest` service to an event bus.
  - `processMessages()`: Turns a batch of SQS messages (returned by `receiveNextMessageBatch()`) into GOST `grant` objects that are inserted/updated into the `grants` table.
- Adds a new Node.js script at `packages/server/src/scripts/consumeGrantModifications.js` that orchestrates the usage of the library, continuously long-polling for and processing any received SQS messages until a shut down is requested (which it attempts to perform in a graceful manner). This script serves as the entrypoint for the work in #1296, which will start ECS tasks that run the script according to the number of messages in the source SQS queue.
- Defines a new helper function, `getSQSClient()`, in the `gost-aws` module that initializes SQS clients from the AWS SDK with conditional configuration for LocalStack. This follows the pattern we've established for SES and S3 client initialization.
- Creates a Knex migration that adds a new `award_floor` column to the `grants` table. A few things to mention here:
  - Although it's used to store dollar values, it's implemented as a text column to stay consistent with the existing `award_ceiling` column. I'm not sure it matters much given our current usage, but we may want to migrate to a more appropriate data type for these columns in the future.  
  - I was surprised to learn that we weren't already persisting this data to a dedicated column; apparently, we just retrieve it from the JSON data stored in the `raw_body` column/field in the following scenarios:
    -  [On the server-side](https://github.com/usdigitalresponse/usdr-gost/blob/b0c8cd447c4b0edbfe0eb6d87b1c8a529f0cb0f3/packages/server/src/routes/grants.js#L13-L33), when [generating CSV exports](https://github.com/usdigitalresponse/usdr-gost/blob/b0c8cd447c4b0edbfe0eb6d87b1c8a529f0cb0f3/packages/server/src/routes/grants.js#L125).
    - [On the client-side](https://github.com/usdigitalresponse/usdr-gost/blob/b0c8cd447c4b0edbfe0eb6d87b1c8a529f0cb0f3/packages/client/src/components/GrantsTable.vue#L282-L302), when [preparing the grants table view](https://github.com/usdigitalresponse/usdr-gost/blob/b0c8cd447c4b0edbfe0eb6d87b1c8a529f0cb0f3/packages/client/src/components/GrantsTable.vue#L186).
  - Once we've started populating the `grants` table from the `grants-ingest` service, we should revisit and replace the usages of `raw_body` (side note: this is the only use-case I can find where we actually read that column) with the new `award_floor` column.
- Misc:
  - Adds tests for the `grants-ingest` module.
  - Adds `@aws-sdk/client-sqs` as a new dependency to the server package.
  - Updated the `localstack/entrypoint/init-aws.sh` script to create an SQS queue named `grants-ingest-events` on startup.
  - Add a new `$GRANTS_INGEST_EVENTS_QUEUE_URL` env var to `packages/server/.env.example` that points to the queue on LocalStack for development environments.

## Screenshots / Demo Video

![938-demo](https://github.com/usdigitalresponse/usdr-gost/assets/1851017/db35fd16-c729-4ce6-84e6-c0ed187f6b1a)

## Testing

**Note:** When testing with docker compose, you can start the script by prefixing the command in Step 3 with `docker compose exec -w /app/packages/server app`. However, it's not possible to observe the graceful shutdown behavior by running the script in this manner since the signal is sent to `docker compose` rather than the script running in the container. Instead, you have to access a shell inside the container by running `docker compose exec app bash` and then running `yarn consume-grants` or `node src/scripts/consumeGrantModifications.js` from the `packages/server` directory. Additionally, to stop the script by issuing `kill -SIGTERM <pid>`, you also need to `docker compose exec app bash` in (yet another) terminal.

1. Ensure that the `grants-ingest-events` SQS queue exists in LocalStack and that the `$GRANTS_INGEST_EVENTS_QUEUE_URL` env var is set to its queue URL.
2. Create a file named something like `sqs-message.json` that contains JSON resembling what we expect to receive from the queue (or some invalid data if you want to test how errors behave). 
    - **Note:** I'll drop an example in a comment on this PR that you can use to test.
4. Start the script by running `yarn consume-grants` from the `server` package or `node /path/to/consumeGrantModifications.js` (I recommend doing this!). 
    - Once it enters the main loop, it should log a message like `Long-polling next SQS message batch from <queue url>`.
5. Send a message to the queue by running `awslocal sqs send-message --queue-url $GRANTS_INGEST_EVENTS_QUEUE_URL --message-body file:///path/to/sqs-message.json`
    - If desired, you can run this command multiple times, either providing the same message or different messages (e.g. invalid JSON or other grants).
7. Observe the logs from the script. If valid JSON is received, it should indicate that it successfully parsed its content and inserted/updated the `grants` table accordingly. Once a batch has been processed, it should go back to logging the same `Long-polling next SQS message batch...` message about every 20 seconds (the configured long-poll duration).
8. At this point, you can keep sending different messages to the queue, or continue to the next step.
9. Check that the database has been updated according to the data you provided to the queue.
10. Request that the script shut down in any of the following ways:
    - Issue a `SIGINT` by pressing `CTRL+C` in the shell where the script is running. 
      - Note that when you do this, your shell will likely go back to a normal prompt but the script will continue to run until the current loop iteration has finished. When it finally stops, you'll get a detached `Shutting down` log message printed to your console.
    - In a separate shell, find the PID for the script (e.g. `ps aux | grep consume`) and issue a `SIGTERM` by running `kill -SIGTERM <pid>`.
      - Note: If you started the script with `yarn` rather than directly with `node`, your shell will probably also detach from the process but continue to print script logs. It's also important to target the PID belonging to the `node` process, rather than the PID associated with the `yarn` command. Frankly, I found it much less cumbersome to run `node /path/to/consumeGrantModifications.js` for this reason, but you do you :)

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
